src/cmd/services/m3coordinator/downsample/downsample_id.go

```diff
@@ -0,0 +1,330 @@
+func (it *encodedTagsIterator) TagValue(tagName []byte) ([]byte, bool) {
+	it.tagDecoder.Reset(it.bytes)
+	for it.Next() {
+		name, value := it.Current()
+		if bytes.Equal(tagName, name) {
+			return value, true
+		}
+	}
+	return nil, false
+}
+
+// Next returns true if there are more tag names and values.
+func (it *encodedTagsIterator) Next() bool {
+	return it.tagDecoder.Next()
+}
+
+// Current returns the current tag name and value.
+func (it *encodedTagsIterator) Current() ([]byte, []byte) {
+	tag := it.tagDecoder.Current()
+	return tag.Name.Bytes(), tag.Value.Bytes()
+}
+
+// Err returns any errors encountered.
+func (it *encodedTagsIterator) Err() error {
+	return it.tagDecoder.Err()
+}
+
+// Close closes the iterator.
+func (it *encodedTagsIterator) Close() {
+	it.bytes.Reset(nil)
+	it.tagDecoder.Reset(it.bytes)
+
+	if it.pool != nil {
+		it.pool.Put(it)
+	}
+}
+
+type encodedTagsIteratorPool struct {
+	tagDecoderPool serialize.TagDecoderPool
+	pool           pool.ObjectPool
+}
+
+func newEncodedTagsIteratorPool(
+	tagDecoderPool serialize.TagDecoderPool,
+	opts pool.ObjectPoolOptions,
+) *encodedTagsIteratorPool {
+	return &encodedTagsIteratorPool{
+		tagDecoderPool: tagDecoderPool,
+		pool:           pool.NewObjectPool(opts),
+	}
+}
+
+func (p *encodedTagsIteratorPool) Init() {
+	p.tagDecoderPool.Init()
+	p.pool.Init(func() interface{} {
+		return newEncodedTagsIterator(p.tagDecoderPool.Get(), p)
+	})
+}
+
+func (p *encodedTagsIteratorPool) Get() *encodedTagsIterator {
+	return p.pool.Get().(*encodedTagsIterator)
+}
+
+func (p *encodedTagsIteratorPool) Put(v *encodedTagsIterator) {
+	p.pool.Put(v)
+}
+
+func isRollupID(
+	tags []byte,
+	iteratorPool *encodedTagsIteratorPool,
+) bool {
+	iter := iteratorPool.Get()
+	iter.Reset(tags)
+
+	for iter.Next() {
+		name, value := iter.Current()
+		if bytes.Equal(name, rollupTagName) && bytes.Equal(value, rollupTagValue) {
+			return true
+		}
+	}
+	iter.Close()
+
+	return false
+}
+
+type rollupIDProvider struct {
+	index          int
+	tagPairs       []id.TagPair
+	rollupTagIndex int
+
+	tagEncoder serialize.TagEncoder
+	pool       *rollupIDProviderPool
+}
+
+func newRollupIDProvider(
+	tagEncoder serialize.TagEncoder,
+	pool *rollupIDProviderPool,
+) *rollupIDProvider {
+	return &rollupIDProvider{
+		tagEncoder: tagEncoder,
+		pool:       pool,
+	}
+}
+
+func (p *rollupIDProvider) provide(
+	tagPairs []id.TagPair,
+) ([]byte, error) {
+	p.reset(tagPairs)
+	p.tagEncoder.Reset()
+	if err := p.tagEncoder.Encode(p); err != nil {
+		return nil, err
+	}
+	data, ok := p.tagEncoder.Data()
+	if !ok {
+		return nil, fmt.Errorf("unable to access encoded tags: ok=%v", ok)
+	}
+	// Need to return a copy
+	return append([]byte(nil), data.Bytes()...), nil
+}
+
+func (p *rollupIDProvider) reset(
+	tagPairs []id.TagPair,
+) {
+	p.index = -1
+	p.tagPairs = tagPairs
+	p.rollupTagIndex = -1
+	for idx, pair := range tagPairs {
+		if bytes.Compare(rollupTagName, pair.Name) < 0 {
```

Comment:
What if there is no rollup tag in the list of tag pairs provided?

Reply:
(As per other comment, there is meant to be always the rollup tag injected so there will never be one in the tag pairs provided as we inject it ourselves)